source/extensions/filters/http/jwt_authn/extractor.cc

```diff
@@ -0,0 +1,154 @@
+#include "extensions/filters/http/jwt_authn/extractor.h"
+
+#include "common/common/utility.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+#include "common/singleton/const_singleton.h"
+
+using ::Envoy::Http::LowerCaseString;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtAuthentication;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+namespace {
+
+/**
+ * Contant values
+ */
+struct JwtConstValueStruct {
+  // The header value prefix for Authorization.
+  const std::string BearerPrefix{"Bearer "};
+
+  // The default query parameter name to extract JWT token
+  const std::string AccessTokenParam{"access_token"};
+};
+typedef ConstSingleton<JwtConstValueStruct> JwtConstValues;
+
+// A base JwtLocation object to store token and specified_issuers.
+class JwtLocationBase : public JwtLocation {
+public:
+  JwtLocationBase(const std::string& token, const std::unordered_set<std::string>& issuers)
+      : token_(token), specified_issuers_(issuers) {}
+
+  // Get the token string
+  const std::string& token() const override { return token_; }
+
+  // Check if an issuer has specified the location.
+  bool isIssuerSpecified(const std::string& issuer) const override {
+    return specified_issuers_.find(issuer) != specified_issuers_.end();
+  }
+
+private:
+  // Extracted token.
+  std::string token_;
+  // Stored issuers specified the location.
+  const std::unordered_set<std::string>& specified_issuers_;
+};
+
+// The JwtLocation for header extraction.
+class JwtHeaderLocation : public JwtLocationBase {
+public:
+  JwtHeaderLocation(const std::string& token, const std::unordered_set<std::string>& issuers,
+                    const LowerCaseString& header)
+      : JwtLocationBase(token, issuers), header_(header) {}
+
+  void removeJwt(Http::HeaderMap& headers) const override { headers.remove(header_); }
+
+private:
+  // the header name the JWT is extracted from.
+  const LowerCaseString& header_;
+};
+
+// The JwtLocation for param extraction.
+class JwtParamLocation : public JwtLocationBase {
+public:
+  JwtParamLocation(const std::string& token, const std::unordered_set<std::string>& issuers,
+                   const std::string&)
+      : JwtLocationBase(token, issuers) {}
+
+  void removeJwt(Http::HeaderMap&) const override {
+    // TODO(qiwzhang): remove JWT from parameter.
+  }
+
+private:
```

Comment:
Are you planning on adding this in this PR?

Reply:
No in this PR.  It is very tricky to remove a query parameter in Path(), very easy to introduce bugs.

Reply:
Done